daemon/commands.c

```diff
@@ -0,0 +1,539 @@
+    *reply_string_size += len;
+}
+
+static void send_command_reply(uv_stream_t *client, cmd_status_t status, char *message)
+{
+    int ret;
+    char reply_string[MAX_COMMAND_LENGTH] = {'\0', };
+    unsigned reply_string_size = 0;
+    uv_buf_t write_buf;
+    uv_write_t write_req;
+
+    add_string_to_command_reply(reply_string, &reply_string_size, cmd_status_str[status]);
+    if (message) {
+        add_string_to_command_reply(reply_string, &reply_string_size, ": ");
+        add_string_to_command_reply(reply_string, &reply_string_size, message);
+    }
+
+    write_req.data = client;
+    write_buf.base = reply_string;
+    write_buf.len = strlen((char *)write_buf.base);
+    ret = uv_write(&write_req, (uv_stream_t *)client, &write_buf, 1, pipe_write_cb);
+    if (ret) {
+        error("uv_write(): %s", uv_strerror(ret));
+    }
+    info("COMMAND: Sending reply: \"%s\"", reply_string);
+}
+
+cmd_status_t execute_command(cmd_t idx, char *args, char **message)
+{
+    cmd_status_t status;
+    cmd_type_t type = command_info_array[idx].type;
+
+    cmd_lock_by_type[type](idx);
+    status = command_info_array[idx].func(args, message);
+    cmd_unlock_by_type[type](idx);
+
+    return status;
+}
+
+static void after_schedule_command(uv_work_t *req, int status)
+{
+    struct command_context *cmd_ctx = req->data;
+
+    (void)status;
+
+    send_command_reply(cmd_ctx->client, cmd_ctx->status, cmd_ctx->message);
+    if (cmd_ctx->message)
+        freez(cmd_ctx->message);
+}
+
+static void schedule_command(uv_work_t *req)
+{
+    struct command_context *cmd_ctx = req->data;
+
+    cmd_ctx->status = execute_command(cmd_ctx->idx, cmd_ctx->args, &cmd_ctx->message);
+}
+
+static void parse_commands(uv_stream_t *client)
+{
+    char *message = NULL, *pos;
+    cmd_t i;
+    cmd_status_t status;
+    struct command_context *cmd_ctx;
+
+    status = CMD_STATUS_FAILURE;
+
+    for (i = 0 ; i < CMD_TOTAL_COMMANDS ; ++i) {
+        pos = strstr(command_string, command_info_array[i].cmd_str);
+        if (pos) {
+            cmd_ctx = mallocz(sizeof(*cmd_ctx));
+            cmd_ctx->work.data = cmd_ctx;
+            cmd_ctx->client = client;
+            cmd_ctx->idx = i;
+            cmd_ctx->args = pos + strlen(command_info_array[i].cmd_str);
+            cmd_ctx->message = NULL;
+
+            assert(0 == uv_queue_work(loop, &cmd_ctx->work, schedule_command, after_schedule_command));
+            break;
+        }
+    }
+    if (CMD_TOTAL_COMMANDS == i) {
+        /* no command found */
+        message = strdupz("Illegal command. Please type \"help\" for instructions.");
+        send_command_reply(client, status, message);
+        freez(message);
+    }
+}
+
+static void pipe_read_cb(uv_stream_t *client, ssize_t nread, const uv_buf_t *buf)
+{
+    if (0 == nread) {
+        info("%s: Zero bytes read by command pipe.", __func__);
+    } else if (UV_EOF == nread) {
+        info("EOF found in command pipe.");
+        parse_commands(client);
```

Comment:
If we only handle commands when the pipe closes then this will work like a batch interface instead of a CLI. Is this the intention?

Reply:
Each invocation of `cli` issue exactly one command.